[JavaScriptCore] Prepare for InstallAPI

emw-apple opened the pull request:
#### 464f082ed06c3b7e262e48d8b6bc49fbc80d06e0
<pre>
[JavaScriptCore] Prepare for InstallAPI
<a href="https://bugs.webkit.org/show_bug.cgi?id=254783">https://bugs.webkit.org/show_bug.cgi?id=254783</a>
rdar://107446500

Reviewed by Mark Lam and Alexey Proskuryakov.

`tapi installapi` is a compiler technology which produces a library stub
by parsing headers declarations. For it to work correctly, symbols
declared in headers need to exactly match the symbols in the binary.
Work towards enabling InstallAPI by fixing up headers in JavaScriptCore.

There are a few common themes to these header fixups:

JavaScriptCore&apos;s private SPI has no umbrella header, so TAPI reads
private headers in lexicographic order. Many headers need additional
includes when imported in this order.

Some headers are architecture-specific and were included contextually.
They need CPU preprocessor guards so that TAPI only parses the right
declarations for the target(s) it is generating.

Until TAPI, there has never been verification that header declarations
are cleaned up when a function implementation is removed. Some
stale declarations can simply be deleted.

* Source/JavaScriptCore/API/JSManagedValueInternal.h: Add missing
  include revealed by TAPI&apos;s include order.
* Source/JavaScriptCore/API/JSValueInternal.h: Add extern &quot;C&quot; fences.
* Source/JavaScriptCore/API/JSWeakObjectMapRefPrivate.cpp: Remove
  undeclared function that was left behind to support nightlies in 2011.
* Source/JavaScriptCore/API/ObjCCallbackFunction.h: Add needed forward
  declaration revealed by TAPI&apos;s include order.
* Source/JavaScriptCore/CMakeLists.txt: Treat LLIntThunks.h as a private
  framework header (as it already is in Xcode).
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
  - Change some header visibilities. Many headers that are generated
    and/or used in context-specific ways, such as look-up tables, are
    not really part of JavaScriptCore&apos;s API and do not export symbols.
    They also tend not to work well with TAPI, because we configure TAPI
    to parse all headers as Objective-C++. Remove these headers from
    &quot;Private&quot; and &quot;Project&quot; membership so that they will not be given to
    TAPI. They are still importable in the project via normal header
    search paths.
  - Add file references for B3CanonicalizePrePostIncrements, which were
    missing, despite B3CanonicalizePrePostIncrements.cpp being compiled
    in unified sources.
  - Add file references for missing headers, AirTmpWidthInlines.h,
    WasmLLIntPlan.h, and WasmEntryPlan.h.
  - Remove reference to DeprecatedInspectorValues.cpp.
* Source/JavaScriptCore/assembler/ARM64Registers.h: Guard with
  CPU(ARM64).
* Source/JavaScriptCore/assembler/ARMv7Registers.h: Guard with
  CPU(ARM_THUMB2).
* Source/JavaScriptCore/assembler/AbstractMacroAssembler.h: Add missing
  includes.
* Source/JavaScriptCore/assembler/AssemblerBuffer.h: Use the same if
  guard on threadSpecificAssemblerHashes() as is used in the
  implementation.
* Source/JavaScriptCore/assembler/FastJITPermissions.h: Add missing
  includes.
* Source/JavaScriptCore/assembler/MIPSRegisters.h: Guard with CPU(MIPS).
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h: Guard with CPU(ARM64).
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h: Guard with
  CPU(ARM_THUMB2).
* Source/JavaScriptCore/assembler/MacroAssemblerX86Common.h: Guard with
  (CPU(X86) || CPU(X86_64)).
* Source/JavaScriptCore/assembler/RISCV64Registers.h: Guard with
  CPU(RISCV64).
* Source/JavaScriptCore/assembler/X86Registers.h: Guard with CPU(X86).
* Source/JavaScriptCore/assembler/X86_64Registers.h: Guard with
  CPU(X86_64).
* Source/JavaScriptCore/b3/B3MemoryValue.h: Mark inlined functions as
  `inline` to avoid declaring symbol visibility.
* Source/JavaScriptCore/b3/B3MemoryValueInlines.h:
(JSC::B3::MemoryValue::isCanonicalWidth const): Move inline
  definition here to prevent TAPI from having to parse it.
* Source/JavaScriptCore/bytecode/AccessCase.h: Add missing include.
* Source/JavaScriptCore/bytecode/AccessCaseSnippetParams.h: Add missing
  include.
* Source/JavaScriptCore/bytecode/RepatchInlines.h: Add missing includes.
* Source/JavaScriptCore/bytecode/Watchpoint.h: Remove declaration for
  deleted fireAllSlow() function.
* Source/JavaScriptCore/disassembler/ARM64/A64DOpcode.h: Add missing
  includes.
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::isInspectorDebuggerAgent const):
  Moved from an inline implementation to match its declaration.
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.h:
  Remove inlined isInspectorDebuggerAgent(). We declare a visible
  symbol, so there must be a non-inline implementation.
* Source/JavaScriptCore/inspector/augmentable/AlternateDispatchableAgent.h:
  Add missing include.
* Source/JavaScriptCore/inspector/cocoa/DeprecatedInspectorValues.cpp:
  Removed. These declarations were removed in 2017, so it is safe to
  remove their implementation now.
* Source/JavaScriptCore/inspector/remote/RemoteInspector.h: Remove
  declaration for setParentProcessInformationIsDelayed() function, which
  was never implemented.
* Source/JavaScriptCore/jit/ExecutableMemoryHandle.h: Add missing
  include.
* Source/JavaScriptCore/jit/JITSizeStatistics.h: Add missing include.
* Source/JavaScriptCore/llint/LLIntOfflineAsmConfig.h: Add a config
  macro for JITCage, needed below.
* Source/JavaScriptCore/llint/LLIntThunks.h: Entry points to CSSJIT were
  visible but undeclared (and were declared in WebCore). Move these
  declarations here so that JSC&apos;s API matches.
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm: The declaration
  for jitCagePtr is guarded by ENABLE(JIT_CAGE), but the implementation
  is unconditionally assembled. Use the JIT_CAGE setting added above to
  conditionally emit it.
* Source/JavaScriptCore/parser/ASTBuilder.h: Add missing include.
* Source/JavaScriptCore/parser/SyntaxChecker.h: Add needed forward
  declaration.
* Source/JavaScriptCore/runtime/CachePayload.h: Remove declaration for
  undefined operator=() function.
* Source/JavaScriptCore/runtime/DeferredWorkTimer.h: Mark inline
  functions as inline, to avoid declaraing symbols with these names.
* Source/JavaScriptCore/runtime/HashMapImpl.h: Remove declarations for
  undefined getHashMapImpl* functions.
* Source/JavaScriptCore/runtime/Identifier.h: Remove declaration for
  undefined add() function.
* Source/JavaScriptCore/runtime/JSCPtrTag.h: Guard pointer tag
  declarations with CPU(ARM64E).
* Source/JavaScriptCore/runtime/JSGlobalObject.h: Remove declaration for
  undefined registerImportMap() function.
* Source/JavaScriptCore/runtime/JSObject.h: Remove declaration for
  undefined setUpStaticFunctionSlot() function.
* Source/JavaScriptCore/runtime/JSTypedArrays.h: Remove declaration for
  undefined createUint8TypedArray() function.
* Source/JavaScriptCore/runtime/ThrowScope.h: Remove declaration for
  undefined printIfNeedCheck() function.
* Source/JavaScriptCore/runtime/VM.cpp: There is a tapi bug which
  prevents it from recognizing the symbol declared from a pure virtual
  destructor (rdar://107384624). Work around this by inlining the
  (empty) destructor. The class is still abstract because of its other
  pure virtual function.
(JSC::VM::ClientData::~ClientData): Deleted.
* Source/JavaScriptCore/runtime/VM.h: Make inlining change described
  above.
(JSC::VM::ClientData::~ClientData):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h: Add missing
  includes.

* Source/WebCore/cssjit/SelectorCompiler.cpp: Adjust namespacing of the CSSJIT
  entry point declarations.
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateReturn):
* Source/WebCore/cssjit/SelectorCompiler.h: Use CSSJIT entry points from
  LLIntThunks.h.
(WebCore::SelectorCompiler::ruleCollectorSimpleSelectorChecker):
(WebCore::SelectorCompiler::querySelectorSimpleSelectorChecker):
(WebCore::SelectorCompiler::ruleCollectorSelectorCheckerWithCheckingContext):
(WebCore::SelectorCompiler::querySelectorSelectorCheckerWithCheckingContext):

Canonical link: <a href="https://commits.webkit.org/262926@main">https://commits.webkit.org/262926@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c95d10822ea8961d527428c0f70bae1d6501fffa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2955 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3019 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3117 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4360 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3361 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2922 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3095 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3058 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2589 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2983 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3355 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2643 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4152 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/851 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2622 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2488 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/2438 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2614 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2673 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3901 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/2798 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3023 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2426 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/3019 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2655 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2625 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/731 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/747 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2642 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/3097 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2844 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/860 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/JavaScriptCore/llint/LLIntOfflineAsmConfig.h:
```diff
@@ -188,6 +188,12 @@
 #endif
 
+#if ENABLE(JIT_CAGE)
+#define OFFLINE_ASM_JIT_CAGE 1
```
Where is this used? The commit message says "below", but I don't see it anywhere else in the diff.

emw-apple replied:
offlineasm has ifdef-like conditional compilation statements, which use tokens that expand to `OFFLINE_ASM_`-prefixed macros in generated code. So `if JIT_CAGE` here is the use site:

https://github.com/WebKit/WebKit/blob/9aa6cb6f001fbedca9258e81d0b3a6858f7040f7/Source/JavaScriptCore/llint/LowLevelInterpreter.asm#L1847-L1854

emw-apple replied:
offlineasm has ifdef-like conditional compilation statements, which use tokens that expand to `OFFLINE_ASM_`-prefixed macros in generated code. So `if JIT_CAGE` here is the use site:

https://github.com/WebKit/WebKit/blob/9aa6cb6f001fbedca9258e81d0b3a6858f7040f7/Source/JavaScriptCore/llint/LowLevelInterpreter.asm#L1847-L1854

aproskuryakov commented on Source/JavaScriptCore/runtime/VM.h:
```diff
@@ -302,7 +302,7 @@ class VM : public ThreadSafeRefCounted<VM>, public DoublyLinkedListNode<VM> {
 
     struct ClientData {
-        JS_EXPORT_PRIVATE virtual ~ClientData() = 0;
+        JS_EXPORT_PRIVATE virtual ~ClientData() { };
```
Perhaps this is the best workaround possible, but this does change the interface - subclasses no longer have to override the destructor.

MenloDorian replied:
I think this is fine.

justinmichaud commented on Source/JavaScriptCore/runtime/ThrowScope.h:
```diff
@@ -65,9 +63,6 @@ class ThrowScope : public ExceptionScope {
 #define DECLARE_THROW_SCOPE(vm__) \
     JSC::ThrowScope((vm__), JSC::ExceptionEventLocation(EXCEPTION_SCOPE_POSITION_FOR_ASAN(vm__), __FUNCTION__, __FILE__, __LINE__))
 
-#define throwScopePrintIfNeedCheck(scope__) \
```
I believe this is a debugging tool. Since it is already #def'd out, do we need to remove it? @MenloDorian would know this for sure, I will ping him

MenloDorian replied:
This is fine to remove.  Thanks.

emw-apple committed:
[JavaScriptCore] Prepare for InstallAPI
https://bugs.webkit.org/show_bug.cgi?id=254783
rdar://107446500

Reviewed by Mark Lam and Alexey Proskuryakov.

`tapi installapi` is a compiler technology which produces a library stub
by parsing headers declarations. For it to work correctly, symbols
declared in headers need to exactly match the symbols in the binary.
Work towards enabling InstallAPI by fixing up headers in JavaScriptCore.

There are a few common themes to these header fixups:

JavaScriptCore's private SPI has no umbrella header, so TAPI reads
private headers in lexicographic order. Many headers need additional
includes when imported in this order.

Some headers are architecture-specific and were included contextually.
They need CPU preprocessor guards so that TAPI only parses the right
declarations for the target(s) it is generating.

Until TAPI, there has never been verification that header declarations
are cleaned up when a function implementation is removed. Some
stale declarations can simply be deleted.

* Source/JavaScriptCore/API/JSManagedValueInternal.h: Add missing
  include revealed by TAPI's include order.
* Source/JavaScriptCore/API/JSValueInternal.h: Add extern "C" fences.
* Source/JavaScriptCore/API/JSWeakObjectMapRefPrivate.cpp: Remove
  undeclared function that was left behind to support nightlies in 2011.
* Source/JavaScriptCore/API/ObjCCallbackFunction.h: Add needed forward
  declaration revealed by TAPI's include order.
* Source/JavaScriptCore/CMakeLists.txt: Treat LLIntThunks.h as a private
  framework header (as it already is in Xcode).
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
  - Change some header visibilities. Many headers that are generated
    and/or used in context-specific ways, such as look-up tables, are
    not really part of JavaScriptCore's API and do not export symbols.
    They also tend not to work well with TAPI, because we configure TAPI
    to parse all headers as Objective-C++. Remove these headers from
    "Private" and "Project" membership so that they will not be given to
    TAPI. They are still importable in the project via normal header
    search paths.
  - Add file references for B3CanonicalizePrePostIncrements, which were
    missing, despite B3CanonicalizePrePostIncrements.cpp being compiled
    in unified sources.
  - Add file references for missing headers, AirTmpWidthInlines.h,
    WasmLLIntPlan.h, and WasmEntryPlan.h.
  - Remove reference to DeprecatedInspectorValues.cpp.
* Source/JavaScriptCore/assembler/ARM64Registers.h: Guard with
  CPU(ARM64).
* Source/JavaScriptCore/assembler/ARMv7Registers.h: Guard with
  CPU(ARM_THUMB2).
* Source/JavaScriptCore/assembler/AbstractMacroAssembler.h: Add missing
  includes.
* Source/JavaScriptCore/assembler/AssemblerBuffer.h: Use the same if
  guard on threadSpecificAssemblerHashes() as is used in the
  implementation.
* Source/JavaScriptCore/assembler/FastJITPermissions.h: Add missing
  includes.
* Source/JavaScriptCore/assembler/MIPSRegisters.h: Guard with CPU(MIPS).
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h: Guard with CPU(ARM64).
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h: Guard with
  CPU(ARM_THUMB2).
* Source/JavaScriptCore/assembler/MacroAssemblerX86Common.h: Guard with
  (CPU(X86) || CPU(X86_64)).
* Source/JavaScriptCore/assembler/RISCV64Registers.h: Guard with
  CPU(RISCV64).
* Source/JavaScriptCore/assembler/X86Registers.h: Guard with CPU(X86).
* Source/JavaScriptCore/assembler/X86_64Registers.h: Guard with
  CPU(X86_64).
* Source/JavaScriptCore/b3/B3MemoryValue.h: Mark inlined functions as
  `inline` to avoid declaring symbol visibility.
* Source/JavaScriptCore/b3/B3MemoryValueInlines.h:
(JSC::B3::MemoryValue::isCanonicalWidth const): Move inline
  definition here to prevent TAPI from having to parse it.
* Source/JavaScriptCore/bytecode/AccessCase.h: Add missing include.
* Source/JavaScriptCore/bytecode/AccessCaseSnippetParams.h: Add missing
  include.
* Source/JavaScriptCore/bytecode/RepatchInlines.h: Add missing includes.
* Source/JavaScriptCore/bytecode/Watchpoint.h: Remove declaration for
  deleted fireAllSlow() function.
* Source/JavaScriptCore/disassembler/ARM64/A64DOpcode.h: Add missing
  includes.
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::isInspectorDebuggerAgent const):
  Moved from an inline implementation to match its declaration.
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.h:
  Remove inlined isInspectorDebuggerAgent(). We declare a visible
  symbol, so there must be a non-inline implementation.
* Source/JavaScriptCore/inspector/augmentable/AlternateDispatchableAgent.h:
  Add missing include.
* Source/JavaScriptCore/inspector/cocoa/DeprecatedInspectorValues.cpp:
  Removed. These declarations were removed in 2017, so it is safe to
  remove their implementation now.
* Source/JavaScriptCore/inspector/remote/RemoteInspector.h: Remove
  declaration for setParentProcessInformationIsDelayed() function, which
  was never implemented.
* Source/JavaScriptCore/jit/ExecutableMemoryHandle.h: Add missing
  include.
* Source/JavaScriptCore/jit/JITSizeStatistics.h: Add missing include.
* Source/JavaScriptCore/llint/LLIntOfflineAsmConfig.h: Add a config
  macro for JITCage, needed below.
* Source/JavaScriptCore/llint/LLIntThunks.h: Entry points to CSSJIT were
  visible but undeclared (and were declared in WebCore). Move these
  declarations here so that JSC's API matches.
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm: The declaration
  for jitCagePtr is guarded by ENABLE(JIT_CAGE), but the implementation
  is unconditionally assembled. Use the JIT_CAGE setting added above to
  conditionally emit it.
* Source/JavaScriptCore/parser/ASTBuilder.h: Add missing include.
* Source/JavaScriptCore/parser/SyntaxChecker.h: Add needed forward
  declaration.
* Source/JavaScriptCore/runtime/CachePayload.h: Remove declaration for
  undefined operator=() function.
* Source/JavaScriptCore/runtime/DeferredWorkTimer.h: Mark inline
  functions as inline, to avoid declaraing symbols with these names.
* Source/JavaScriptCore/runtime/HashMapImpl.h: Remove declarations for
  undefined getHashMapImpl* functions.
* Source/JavaScriptCore/runtime/Identifier.h: Remove declaration for
  undefined add() function.
* Source/JavaScriptCore/runtime/JSCPtrTag.h: Guard pointer tag
  declarations with CPU(ARM64E).
* Source/JavaScriptCore/runtime/JSGlobalObject.h: Remove declaration for
  undefined registerImportMap() function.
* Source/JavaScriptCore/runtime/JSObject.h: Remove declaration for
  undefined setUpStaticFunctionSlot() function.
* Source/JavaScriptCore/runtime/JSTypedArrays.h: Remove declaration for
  undefined createUint8TypedArray() function.
* Source/JavaScriptCore/runtime/ThrowScope.h: Remove declaration for
  undefined printIfNeedCheck() function.
* Source/JavaScriptCore/runtime/VM.cpp: There is a tapi bug which
  prevents it from recognizing the symbol declared from a pure virtual
  destructor (rdar://107384624). Work around this by inlining the
  (empty) destructor. The class is still abstract because of its other
  pure virtual function.
(JSC::VM::ClientData::~ClientData): Deleted.
* Source/JavaScriptCore/runtime/VM.h: Make inlining change described
  above.
(JSC::VM::ClientData::~ClientData):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h: Add missing
  includes.

* Source/WebCore/cssjit/SelectorCompiler.cpp: Adjust namespacing of the CSSJIT
  entry point declarations.
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateReturn):
* Source/WebCore/cssjit/SelectorCompiler.h: Use CSSJIT entry points from
  LLIntThunks.h.
(WebCore::SelectorCompiler::ruleCollectorSimpleSelectorChecker):
(WebCore::SelectorCompiler::querySelectorSimpleSelectorChecker):
(WebCore::SelectorCompiler::ruleCollectorSelectorCheckerWithCheckingContext):
(WebCore::SelectorCompiler::querySelectorSelectorCheckerWithCheckingContext):

Canonical link: https://commits.webkit.org/262926@main
